### PR TITLE
adding disposable, shared controller interface and gpiocontroller interface

### PR DESCRIPTION
**Pull request by colombod**
With this PR I am trying to propose a discussion around some interfaces used during the development of the piTop bindings

- introduction of a IGpioController interface to make it possible to have mocks and implement shared proxy
- share controller class to create a view over a controller so that devices and other part of applications can manage pins and callbacks with the ability to clean up just the portion of state interested
- disposable factory to create disposable object to simplify the closing of pins and the unregistration of callbacks

### CODE REVIEW

**Comment by pgrawehr** [comment]
Unfortunatelly, there's a big issue: System.Device.Gpio interfaces must be kept backwards compatible. Breaking changes must have _very_ good reasoning. Note also that some interface concept discussions have been open for a long time already (but where postponed due to other tasks and long absences of the responsible maintainers). See #1128 and #878.

**Comment by colombod** [comment]
> Unfortunatelly, there's a big issue: System.Device.Gpio interfaces must be kept backwards compatible. Breaking changes must have _very_ good reasoning. Note also that some interface concept discussions have been open for a long time already (but where postponed due to other tasks and long absences of the responsible maintainers). See #1128 and #878.

Makes sense, the contract change on the OpenPin and RegisterCallBack is something that could be left out, would it still make the case of having IGpioController interface a big issue with the api backward compatibility?

**Comment by colombod** [comment]
> Unfortunatelly, there's a big issue: System.Device.Gpio interfaces must be kept backwards compatible. Breaking changes must have _very_ good reasoning. Note also that some interface concept discussions have been open for a long time already (but where postponed due to other tasks and long absences of the responsible maintainers). See #1128 and #878.

Thank you for you feedback, I realized I really just need the IGpioController interface so I can build proxy and adaptors, the other api changes can be done as extension methods so not to make the type too big or break existing contract. Thank you

**Comment by pgrawehr** [comment]
> 
> Makes sense, the contract change on the OpenPin and RegisterCallBack is something that could be left out, would it still make the case of having IGpioController interface a big issue with the api backward compatibility?

I guess so. Interfaces don't scale well (because _every_ change to an interface, including adding methods, is a breaking change). What we're probably going to do (is under discussion now), is to unseal the GpioController class and change its methods to virtual. This (besides other reasons) also allow the mocking support you want.

**Comment by colombod** [comment]
@pgrawehr the unselaed class would be fine. Any time frame idea for such a change? That would actually enable all api addition I am doing to be an external package that user can adopt if they are interested in some additional pattern and utilities.

**Comment by pgrawehr** [comment]
@colombod : Vote for/have a look at #1128. That contains quite a few concept updates, and it would be great to have a broad discussion on them.

**Comment by colombod** [comment]
> @colombod : Vote for/have a look at #1128. That contains quite a few concept updates, and it would be great to have a broad discussion on them.

that PR is what I call Christmas present. It is absolutely brilliant with the controller and spi management and all the rest.

**Comment by krwq** [comment]
This seem to overlap with some of the other conversations we have in other issues:
https://github.com/dotnet/iot/issues/383
https://github.com/dotnet/iot/issues/80
https://github.com/dotnet/iot/issues/1182
https://github.com/dotnet/iot/issues/637
https://github.com/dotnet/iot/issues/878

I'm hoping we will figure out the abstraction subject as we're having request from different directions about it

File: src/System.Device.Gpio/System/Device/Gpio/IGpioControllerExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    /// </summary>
+    public static class IGpioControllerExtensions
+    {
+        /// <summary>
+        /// Opens a pin in order for it to be ready to use.
+        /// </summary>
+        /// <param name="controller">The GPioController.</param>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <returns>A disposable that will close the pin if disposed.</returns>
+        public static IDisposable OpenPinAsDisposable(this IGpioController controller, int pinNumber)
```
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/IGpioControllerExtensions.cs:20]: we're actually thinking of having GpioPin class instead, see https://github.com/dotnet/iot/issues/215

File: src/System.Device.Gpio/System/Device/Gpio/IGpioControllerExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+        /// <summary>
+        /// Adds a callback that will be invoked when pinNumber has an event of type eventType.
+        /// </summary>
+        /// <param name="controller">The GPioController.</param>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <param name="eventTypes">The event types to wait for.</param>
+        /// <param name="callback">The callback method that will be invoked.</param>
+        /// <returns>A disposable object that will remove the added callback </returns>
+        public static IDisposable RegisterCallbackForPinValueChangedEventAsDisposable(this IGpioController controller, int pinNumber,
```
Comment (krwq) [inline src/System.Device.Gpio/System/Device/Gpio/IGpioControllerExtensions.cs:47]: I'd rather have us introduce the real abstraction for this (i.e. CallbackForPinValueChanged) implementing IDisposable

**Comment by krwq** [comment]
I do like confirmation from here that adding GpioPin is the right direction and also looking at this PR seems like another person trying to bypass GpioDriver abstraction which makes me thinking to perhaps just deprecate it and make everything implement  GpioController directly (so unseal and virtualize it as proposed by @pgrawehr as well and just not have GpioDriver - and of course leave compat GpioController to make old code still work)

**Comment by pgrawehr** [comment]
> 
> 
> ...and just not have GpioDriver - and of course leave compat GpioController to make old code still work)

I think that extra driver class is fine to keep, even if the Controller is unsealed - it separates responsibilities. The Controller handles pin management, while the driver handles the low-level operations on individual pins.

**Comment by Ellerbach** [comment]
> I do like confirmation from here that adding GpioPin is the right direction

Yes, this really help to clearly see the need of abstraction, separating drivers from the controller itself and having GpioPin as well.

> The Controller handles pin management, while the driver handles the low-level operations on individual pins.

I agree, this would perfectly work. 

> This seem to overlap with some of the other conversations we have in other issues:
> #383
> #80
> #1182
> #637
> #878

And the good news is that with all this, I feel we're all ready to move forward. Once the new model will be present, the trendy topic shouldn't be GpioController anymore on the various issues :-)

**Comment by joperezr** [comment]
[Triage] Hey @colombod we have been making good progress on this PR: https://github.com/dotnet/iot/pull/1254 which might help you out with your needs and might be related to this one. Can you check that one to see if that would solve your needs?

**Comment by colombod** [comment]
> [Triage] Hey @colombod we have been making good progress on this PR: #1254 which might help you out with your needs and might be related to this one. Can you check that one to see if that would solve your needs?

This pr looks good and enables quite few things for me

**Comment by joperezr** [comment]
[Triage] @colombod do you think there is still value on keeping this PR open or would #1254 cover your needs? If the latter then we can go ahead and close this PR.

**Comment by colombod** [comment]
Will close this or as I have different things based on your nee library

**Pull request closed without merging**